### PR TITLE
[OUR415-210] React 16 => 18

### PR DESCRIPTION
**Pull request by rosschapman**
If only we could bump dependencies and everything would _Just Work™️_. Most the app code was spared except for the Algolia components since the underlying apis changed dramatically. For posterity, here is my best attempt at a meaningful list of changes that would be of interest to other devs (in no particular order): 

1. Updates application instantiation to new `createRoot` api
1. Hot module reloading comes for free with webpack 4+, so `react-hot-loader/babel` was removed
1. Search-related globals like `SiteSearchProvider` were removed in favor of leveraging state managed by the top level `InstantSearch` provider with hooks like `useInstantSearch`
1. Added a `setupTests.js` file to boot all tests with the common imports like `@testing-library/jest-dom`
1. Swaps Algolia HOC _connectors_ for their equivalent hooks (this does require rethinking the user interaction handling in some cases)
1. Significantly refactors refinement list components using new hooks
1. Renames refinement list components to more easily identify as belonging to the _Browse_ or _Search_ experience
1. Unwraps the `InnerSearchResults` components which didn't seem to be doing anything meaningful as so-called _presentation_ components
1. Refactors to use the Algolia `<SearchBox/>` component and removes the shadow component. I was having some trouble with the shadow component after update. This seemed like a tough workaround to maintain. We should work toward moving the `<InstantSearch />` and configuration providers up scope to wrap the entire app on routes it's needed. 
1. Removes extra ceremony managing the search state in favor of the new `routing` param on the `InstantSearch`. Overall I am trying to rely on the default behavior of the algolia components as a baseline until we have specific need for customization. Currently the only important customization we need is the eligibility mapping on the search pages. 

One other thing to note is a new doc (`docs/search-requirements.md`) that records the implementation contract for the search behaviors we want to support from the code's POV.

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -11,8 +11,5 @@
     "@babel/preset-react",
     "@babel/preset-typescript"
   ],
-  "plugins": [
-    "@babel/plugin-proposal-class-properties",
-    "react-hot-loader/babel"
```
Comment (rosschapman) [inline .babelrc:16]: HMR (hot module replacement) is handled by webpack now (since v4).

File: app/components/listing/ServiceCard.test.tsx [hunk 2]
```diff
@@ -1,6 +1,5 @@
 import React from "react";
 import { render, screen } from "@testing-library/react";
-import "@testing-library/jest-dom";
```
Comment (rosschapman) [inline app/components/listing/ServiceCard.test.tsx:3]: Moved to newly created `setupTests.js`

File: tsconfig.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     "baseUrl": "app",
-    "lib": ["dom", "dom.iterable", "es2015"],
+    "lib": ["dom", "dom.iterable", "es2021"],
```
Comment (rosschapman) [inline tsconfig.json:4]: 💥 Look at us joining this decade!

**Review by adriencyberspace — APPROVED** [review]
LGTM! Note that we will QA and make new tickets for bugs that arise

**Comment by rosschapman** [comment]
@adriencyberspace and I 🍐 reviewed this together today. There are some bugs with refinement selection. I'm going to work those out in separate commits.

**Pull request merged**